src/io/output_element.hh

```diff
@@ -0,0 +1,290 @@
+{
+    unsigned int n = n_nodes();
+    ASSERT_DBG(loc_idx < n);
+    unsigned int con_off = (*output_mesh_->offsets_)[ele_idx_];
+    return (*connectivity)[con_off - n + loc_idx];
+}
+
+inline OutputElement::Point OutputElement::vertex_internal(unsigned int loc_idx,
+                                                 shared_ptr< MeshData< unsigned int > > connectivity,
+                                                 shared_ptr< MeshData< double > > nodes) const
+{
+    unsigned int n = n_nodes();
+    ASSERT_DBG(loc_idx < n);
+    unsigned int con_off = (*output_mesh_->offsets_)[ele_idx_];
+    unsigned int off = spacedim * (*connectivity)[con_off - n + loc_idx];
+    auto &d = nodes->data_;
+    Point point({d[off], d[off+1], d[off+2]});
+    return point;
+}
+
+inline std::vector< OutputElement::Point > OutputElement::vertex_list_internal(shared_ptr< MeshData< unsigned int > > connectivity, 
+                                                                               shared_ptr< MeshData< double > > nodes) const
+{
+    const unsigned int n = n_nodes();
+    std::vector<Point> vertices(n);
+    
+    unsigned int con_off = (*output_mesh_->offsets_)[ele_idx_];
+    auto &d = nodes->data_;
+    for(unsigned int i=0; i<n; i++) {
+        unsigned int off = spacedim * (*connectivity)[con_off - n + i];
+        vertices[i] = {d[off], d[off+1], d[off+2]};
+        off += spacedim;
+    }
+    return vertices;
+}
+
+
+inline unsigned int OutputElement::node_index(unsigned int loc_idx) const
+{
+    return node_index_internal(loc_idx, output_mesh_->connectivity_);
+}
+
+inline unsigned int OutputElement::node_index_disc(unsigned int loc_idx) const
+{
+    ASSERT_DBG(output_mesh_->discont_data_computed_);
+    return node_index_internal(loc_idx, output_mesh_->discont_connectivity_);
+}
+
+inline OutputElement::Point OutputElement::vertex(unsigned int loc_idx) const
+{
+    return vertex_internal(loc_idx, output_mesh_->connectivity_, output_mesh_->nodes_);
+}
+
+inline OutputElement::Point OutputElement::vertex_disc(unsigned int loc_idx) const
+{
+    return vertex_internal(loc_idx, output_mesh_->discont_connectivity_, output_mesh_->discont_nodes_);
+}
+
+inline std::vector< OutputElement::Point > OutputElement::vertex_list() const
+{
+    return vertex_list_internal(output_mesh_->connectivity_, output_mesh_->nodes_);
+}
+
+inline std::vector< OutputElement::Point > OutputElement::vertex_list_disc() const
+{
+    return vertex_list_internal(output_mesh_->discont_connectivity_, output_mesh_->discont_nodes_);
+}
+
+inline OutputElement::Point OutputElement::centre() const
+{
+    Point res({0,0,0});
+    for(auto& v : vertex_list() ) res += v;
+    return res/n_nodes();
+}
+
+// --------------------------------------------------- OutputElementIterator ---------------------------------
+/** @brief Output element iterator.
+ * Provides iterator over elements of the output mesh.
+ */
+class OutputElementIterator
```

Comment:
The template parameter would then be required to implement a function that increases index (in our case inc() ) in operator++. Should not then the template parameter itself implement operator++?

Secondly, couldn't the implementation of operator ==, be general:
*this == ref_to_other_object, or is it necessary to be able to implement operator== differently for each usage?